core/include/vecmem/memory/memory_resource.hpp

```diff
@@ -7,25 +7,34 @@
 
 #pragma once
 
+// This header can only be used in C++17 mode. So "device compilers" that
+// can't use C++17, must not see it. Which in practically all cases should
+// "just" be a question of code organisation.
+#if __cplusplus < 201700L
+#error \
+    "This header can only be used in C++17 mode. " \
+         "Ideally it should only be used by the \"host compiler\"."
+#endif  // < C++17
+
```

Comment:
I still don't really understand why we have this. If a compiler supports memory resources for C++14 then why shouldn't we allow that to work? It seems a little ham-fisted to me to check for this explicitly, but in the end it's not a big deal.

Reply:
What we check during the CMake configuration in the new version is that the host compiler, with whatever c\+\+ standard it uses, would find these headers/types.

But at that point we may still mistakenly include this header in a CUDA source file, with CUDA set to cpp14. In which case, without this explicit check, they will just get an error that the header is missing. This message is meant to give users a hint, that they were not supposed to include this file in the place where they did.

At least this was my thinking with this...

Reply:
That makes sense, but wouldn't it be more robust, in that case, to either check only against `__CUDACC__`, or check against both `__CUDACC__` _and_ `__cplusplus`?

Reply:
I definitely do not want to check for `__CUDACC__` here. What would be the logic anyway?

If CUDA with C\+\+17 doesn't manage to find the same version of these types as the host compiler (which would be very weird btw, telling us that CUDA is not being used correctly), then getting the error message about a missing header file should be "good enough".

Reply:
Okay, works for me. Let's leave it as it is then. :slightly_smiling_face: